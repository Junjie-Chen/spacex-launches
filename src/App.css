.timeline-header {
  padding-left: 0;
}

.file .timeline-icon.failure {
  background-color: #ea4335;
}

.file .timeline-icon.success {
  background-color: #4285f4;
}
