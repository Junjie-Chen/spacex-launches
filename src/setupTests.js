/**
 * @file Initializes test environment.
 * @author Junjie Chen <junjie.chen18@gmail.com>
 * @see {@link https://create-react-app.dev/docs/running-tests/#initializing-test-environment | Initializing Test Environment}
 */

import { mockData } from './mockData';

export * from './mockData';
export { sortCollection, handleError, cloneDeep } from './utilities';

/**
 * Mock the Fetch API.
 * @returns {Promise} A Promise that is resolved with mock data
 */
global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve(mockData)
  })
);
