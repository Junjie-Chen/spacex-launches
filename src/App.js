import React from 'react';
import './App.css';
import { useEffect, useState } from 'react';
import { sortCollection, handleError } from './utilities';

function fetchLaunches() {
  const [state, setState] = useState({ loading: true });

  useEffect(() => {
    fetch('http://localhost:8080/launches')
      .then(response => response.json())
      .then(data => {
        setState({ data, loading: false });
      })
      .catch(err => {
        console.error(err);
      });
  }, []);

  return state;
}

function getSortedLaunchesByDate(dataCollection, func, sortingFunc) {
  let sortedLaunchesByDate = [];

  const hasError = handleError(() => {
    return sortCollection(dataCollection, func, sortingFunc);
  });

  // TODO: returns the error message if any

  if (hasError.result) {
    sortedLaunchesByDate = hasError.result;
  }

  return sortedLaunchesByDate;
}

function Header() {
  return (
    <div className="page-head">
      <h2 className="page-head-title text-center">Space X Launches</h2>
    </div>
  );
}

function Loading() {
  return (
    <div className="progress">
      <div
        className="progress-bar bg-primary progress-bar-striped progress-bar-animated"
        role="progressbar"
        style={{ width: '100%' }}
        aria-valuenow="100"
        aria-valuemin="0"
        aria-valuemax="100"
      >
        Loading
      </div>
    </div>
  );
}

function Launches({ launches }) {
  const launchesByYear = launches.reduce((list, launch) => {
    const date = launch.launch_date_utc.slice(0, 4);
    list[date] = list[date] || [];
    list[date].push(launch);
    return list;
  }, {});

  let times = 0;

  return (
    <ul data-testid="launches" className="timeline timeline-variant">
      {Object.keys(launchesByYear).map(launchYear => {
        const sortedLaunchesByDate = getSortedLaunchesByDate(
          launchesByYear[launchYear],
          item => {
            if (!item.launch_date_utc) {
              throw new ReferenceError('The date cannot be found! Please check the corresponding property for the date.');
            }

            return [+new Date(item.launch_date_utc), item];
          },
          (a, b) => a - b
        );

        return (
          <span key={launchYear}>
            <li className="timeline-month">{launchYear}</li>
            {sortedLaunchesByDate.map(launch => (
              <Launch key={launch.id} launch={launch} times={++times} />
            ))}
          </span>
        );
      })}
    </ul>
  );
}

function Launch({ launch, times }) {
  const launchAlignment = times % 2 !== 0 ? 'right' : 'left';
  const launchStatus = launch.launch_success ? 'success' : 'failure';
  const launchIcon = launch.launch_success ? (
    <i className="icon mdi mdi-rocket" />
  ) : (
    <i className="icon mdi mdi-bomb" />
  );

  return (
    <li data-testid="launch" className={`timeline-item timeline-item-detailed ${launchAlignment}`}>
      <div className="timeline-content timeline-type file">
        <div data-testid="launch-icon" className={`timeline-icon ${launchStatus}`}>{launchIcon}</div>

        <div className="timeline-header">
          <span data-testid="launch-mission" className="timeline-autor">
            #{launch.id}: {launch.mission_name}
          </span>{' '}
          <p className="timeline-activity">
            {launch.rocket.rocket_name} &mdash; {launch.launch_site.site_name}
          </p>
          <span data-testid="launch-date" className="timeline-time">{launch.launch_date_utc.slice(0, 10)}</span>
        </div>
        <div className="timeline-summary">
          <p>{launch.details}</p>
        </div>
      </div>
    </li>
  );
}

export default function App() {
  const { data, loading } = fetchLaunches();

  return (
    <div>
      <Header />
      {loading ? <Loading /> : <Launches launches={data.launches} />}
    </div>
  );
}
