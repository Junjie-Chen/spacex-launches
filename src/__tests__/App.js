import React from 'react';
import { mockData, sortCollection, cloneDeep } from '../setupTests';
import App from '../App';
import { render, cleanup, waitForElement } from 'react-testing-library';
// TODO: resolve act warning https://github.com/kentcdodds/react-testing-library/issues/281
import { act } from 'react-dom/test-utils';

let component;

beforeEach(async () => {
  /* When testing, code that causes React state updates should be wrapped into act(...):

     act(() => {
       // fire events that update state
     });
     // assert on the output

     This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act */
  await act(async () => {
    component = render(<App />);
    await waitForElement(() => component.getByTestId('launches'));
  });
});

afterEach(cleanup);

test('renders launches', () => {
  expect(component).toMatchSnapshot();
});

test('renders launches in chronological order', () => {
  const renderedLaunchDates = component.getAllByTestId('launch-date');
  const launchesDeepClone = cloneDeep(mockData.launches);
  const sortedLaunchesDeepClone = sortCollection(
    launchesDeepClone,
    item => {
      if (!item.launch_date_utc) {
        throw new ReferenceError('The date cannot be found! Please check the corresponding property for the date.');
      }

      return [+new Date(item.launch_date_utc), item];
    },
    (a, b) => a - b
  );
  const launchDates = sortedLaunchesDeepClone.map(launch => {
    return launch.launch_date_utc.slice(0, 10);
  });

  expect(renderedLaunchDates).toHaveLength(2);
  renderedLaunchDates.forEach((launchDate, index) => {
    expect(launchDate.textContent).toBe(launchDates[index]);
  });
});

test('renders launches with ID', () => {
  const renderedLaunchMissions = component.getAllByTestId('launch-mission');
  const launchesDeepClone = cloneDeep(mockData.launches);
  const sortedLaunchesDeepClone = sortCollection(
    launchesDeepClone,
    (item) => {
      if (!item.launch_date_utc) {
        throw new ReferenceError('The date cannot be found! Please check the corresponding property for the date.');
      }

      return [+new Date(item.launch_date_utc), item];
    },
    (a, b) => a - b
  );
  const launchIDs = sortedLaunchesDeepClone.map(launch => {
    return launch.id;
  });

  expect(renderedLaunchMissions).toHaveLength(2);
  renderedLaunchMissions.forEach((launchMission, index) => {
    expect(launchMission.textContent.split(':')[0].replace(/#/, '')).toBe(launchIDs[index]);
  });
});

test('renders launches in alternate timeline', () => {
  const renderedLaunches = component.getAllByTestId('launch');
  const launchAlignments = ['right', 'left'];

  expect(renderedLaunches).toHaveLength(2);
  renderedLaunches.forEach((launch, index) => {
    expect(launch.classList.contains(launchAlignments[index])).toBe(true);
  });
});

test('renders launch icons in corresponding color', () => {
  const renderedLaunchIcons = component.getAllByTestId('launch-icon');
  const launchStatuses = ['failure', 'success'];

  expect(renderedLaunchIcons).toHaveLength(2);
  renderedLaunchIcons.forEach((launchIcon, index) => {
    expect(launchIcon.classList.contains(launchStatuses[index])).toBe(true);
  });
});
