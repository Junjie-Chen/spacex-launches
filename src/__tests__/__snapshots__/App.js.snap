// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders launches 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="page-head"
        >
          <h2
            class="page-head-title text-center"
          >
            Space X Launches
          </h2>
        </div>
        <ul
          class="timeline timeline-variant"
          data-testid="launches"
        >
          <span>
            <li
              class="timeline-month"
            >
              2018
            </li>
            <li
              class="timeline-item timeline-item-detailed right"
              data-testid="launch"
            >
              <div
                class="timeline-content timeline-type file"
              >
                <div
                  class="timeline-icon failure"
                  data-testid="launch-icon"
                >
                  <i
                    class="icon mdi mdi-bomb"
                  />
                </div>
                <div
                  class="timeline-header"
                >
                  <span
                    class="timeline-autor"
                    data-testid="launch-mission"
                  >
                    #
                    54
                    : 
                    SES-16 / GovSat-1
                  </span>
                   
                  <p
                    class="timeline-activity"
                  >
                    Falcon 9
                     — 
                    CCAFS SLC 40
                  </p>
                  <span
                    class="timeline-time"
                    data-testid="launch-date"
                  >
                    2018-01-31
                  </span>
                </div>
                <div
                  class="timeline-summary"
                >
                  <p>
                    Reused booster from the classified NROL-76 mission in May 2017. Following a successful experimental ocean landing that used three engines, the booster unexpectedly remained intact; Elon Musk stated in a tweet that SpaceX will attempt to tow the booster to shore.
                  </p>
                </div>
              </div>
            </li>
            <li
              class="timeline-item timeline-item-detailed left"
              data-testid="launch"
            >
              <div
                class="timeline-content timeline-type file"
              >
                <div
                  class="timeline-icon success"
                  data-testid="launch-icon"
                >
                  <i
                    class="icon mdi mdi-rocket"
                  />
                </div>
                <div
                  class="timeline-header"
                >
                  <span
                    class="timeline-autor"
                    data-testid="launch-mission"
                  >
                    #
                    57
                    : 
                    Hispasat 30W-6
                  </span>
                   
                  <p
                    class="timeline-activity"
                  >
                    Falcon 9
                     — 
                    CCAFS SLC 40
                  </p>
                  <span
                    class="timeline-time"
                    data-testid="launch-date"
                  >
                    2018-03-06
                  </span>
                </div>
                <div
                  class="timeline-summary"
                >
                  <p>
                    Launched with landing legs and titanium grid fins. Did not attempt a landing due to 'unfavorable weather conditions in the recovery area'.
                  </p>
                </div>
              </div>
            </li>
          </span>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="page-head"
      >
        <h2
          class="page-head-title text-center"
        >
          Space X Launches
        </h2>
      </div>
      <ul
        class="timeline timeline-variant"
        data-testid="launches"
      >
        <span>
          <li
            class="timeline-month"
          >
            2018
          </li>
          <li
            class="timeline-item timeline-item-detailed right"
            data-testid="launch"
          >
            <div
              class="timeline-content timeline-type file"
            >
              <div
                class="timeline-icon failure"
                data-testid="launch-icon"
              >
                <i
                  class="icon mdi mdi-bomb"
                />
              </div>
              <div
                class="timeline-header"
              >
                <span
                  class="timeline-autor"
                  data-testid="launch-mission"
                >
                  #
                  54
                  : 
                  SES-16 / GovSat-1
                </span>
                 
                <p
                  class="timeline-activity"
                >
                  Falcon 9
                   — 
                  CCAFS SLC 40
                </p>
                <span
                  class="timeline-time"
                  data-testid="launch-date"
                >
                  2018-01-31
                </span>
              </div>
              <div
                class="timeline-summary"
              >
                <p>
                  Reused booster from the classified NROL-76 mission in May 2017. Following a successful experimental ocean landing that used three engines, the booster unexpectedly remained intact; Elon Musk stated in a tweet that SpaceX will attempt to tow the booster to shore.
                </p>
              </div>
            </div>
          </li>
          <li
            class="timeline-item timeline-item-detailed left"
            data-testid="launch"
          >
            <div
              class="timeline-content timeline-type file"
            >
              <div
                class="timeline-icon success"
                data-testid="launch-icon"
              >
                <i
                  class="icon mdi mdi-rocket"
                />
              </div>
              <div
                class="timeline-header"
              >
                <span
                  class="timeline-autor"
                  data-testid="launch-mission"
                >
                  #
                  57
                  : 
                  Hispasat 30W-6
                </span>
                 
                <p
                  class="timeline-activity"
                >
                  Falcon 9
                   — 
                  CCAFS SLC 40
                </p>
                <span
                  class="timeline-time"
                  data-testid="launch-date"
                >
                  2018-03-06
                </span>
              </div>
              <div
                class="timeline-summary"
              >
                <p>
                  Launched with landing legs and titanium grid fins. Did not attempt a landing due to 'unfavorable weather conditions in the recovery area'.
                </p>
              </div>
            </div>
          </li>
        </span>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
